### PR TITLE
controller: export wallclock lag metrics also for storage collections

### PR DESCRIPTION
**Pull request by teskje**
This PR rewires the existing `mz_dataflow_wallclock_lag_seconds` metric so that it also includes storage collections. To this end, a new `ControllerMetrics` type is introduced to define metrics that are exported by both the compute and the storage controller, and the wallclock lag metrics are moved there. The `ControllerMetrics` type is then passed to both controllers, so they can export wallclock lag metrics for their respective collections.

Note that in contrast to compute collections, the wallclock lag for storage collections is not per replica (as we are comparing with the global persist frontier here), and in some cases not even per cluster (as not all storage collections are associated with clusters). As a result, the `replica_id` label is always empty for storage collections, and the `instance_id` label is sometimes empty.

### Motivation

  * This PR adds a known-desirable feature.

Part of https://github.com/MaterializeInc/database-issues/issues/8235

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/cluster-client/src/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    pub fn wallclock_lag_metrics(
+        &self,
+        collection_id: String,
+        instance_id: Option<String>,
+        replica_id: Option<String>,
+    ) -> WallclockLagMetrics {
+        let labels = vec![
+            instance_id.unwrap_or_default(),
+            replica_id.unwrap_or_default(),
+            collection_id.to_string(),
```
Comment (petrosagg) [inline src/cluster-client/src/metrics.rs:66]: Isn't this already a string?

File: src/cluster-client/src/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    }
+}
+
+/// Metrics tracking frontier wallclock lag for a collection.
+#[derive(Debug)]
+pub struct WallclockLagMetrics {
+    /// Gauge tracking minimum dataflow wallclock lag.
+    wallclock_lag_seconds_min: DeleteOnDropGauge<'static, AtomicF64, Vec<String>>,
+    /// Gauge tracking maximum dataflow wallclock lag.
+    wallclock_lag_seconds_max: DeleteOnDropGauge<'static, AtomicF64, Vec<String>>,
```
Comment (petrosagg) [inline src/cluster-client/src/metrics.rs:107]: I was surprised that these are floats. Do we ever expect non-integer values for these calculations?

File: src/cluster-client/src/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    /// Counter tracking the total sum of dataflow wallclock lag.
+    wallclock_lag_seconds_sum: DeleteOnDropCounter<'static, AtomicF64, Vec<String>>,
+    /// Counter tracking the total count of dataflow wallclock lag measurements.
+    wallclock_lag_seconds_count: DeleteOnDropCounter<'static, AtomicU64, Vec<String>>,
+
+    /// State maintaining minimum and maximum wallclock lag.
+    wallclock_lag_minmax: SlidingMinMax<f32>,
```
Comment (petrosagg) [inline src/cluster-client/src/metrics.rs:114]: Also a bit confused here, if we're storing F64 above why do we use f32 here?

File: src/cluster-client/src/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+}
+
+impl WallclockLagMetrics {
+    /// Observe a new wallclock lag measurement.
+    pub fn observe(&mut self, lag: Duration) {
+        let lag_secs = lag.as_secs_f32();
+
+        self.wallclock_lag_minmax.add_sample(lag_secs);
```
Comment (petrosagg) [inline src/cluster-client/src/metrics.rs:122]: To my untrained eye this state seems to be storing exactly the same info that `wallclock_lag_seconds_{min,max}` are storing. Why do we need to store them twice?

File: src/cluster-client/src/metrics.rs (hunk 1 above)
Comment (teskje) [reply src/cluster-client/src/metrics.rs:66]: Yes, not sure what I was thinking!

File: src/cluster-client/src/metrics.rs (hunk 3 above)
Comment (teskje) [reply src/cluster-client/src/metrics.rs:114]: To safe memory! There is some more discussion in the comments [here](https://github.com/MaterializeInc/materialize/pull/29140), but the gist is that it reduces the memory required for tracking this metric from ~1.5KB to half that, for each collection. There can be a lot of collections in an environment, so the difference matters.

File: src/cluster-client/src/metrics.rs (hunk 4 above)
Comment (teskje) [reply src/cluster-client/src/metrics.rs:122]: `wallclock_lag_minmax` is the sliding min/max over the last minute (really: over the last 60 samples, where we mean to collect a sample every second). Whereas `wallclock_lag_seconds_{min,max}` are the Prometheus metrics that only store a single value, they cannot be used to calculate a sliding window. We want to Prometheus metrics to always reflect the min/max over the last minute, and we maintain `wallclock_lag_minmax` to achieve that.

**Review by petrosagg — APPROVED** [review]
Thanks for all the explanations! Code looks good to me!

**Comment by teskje** [comment]
TFTR!

**Pull request merged**